### PR TITLE
chore:  optionally use emptyDir for wal

### LINKED ISSUE #449: reduce should use a persistent store by default
Today the default store is an in-memory store. Data will be lost during pod migrations and restarts. We should use WAL by default and can give the option to override to in-memory store if required (not recommended though).

### PR DESCRIPTION
**Pull request by yhl25**
Signed-off-by: Yashash H L <yashashhl25@gmail.com>

Fixes #449

### CODE REVIEW

File: pkg/reconciler/vertex/controller.go [hunk 1]
```diff
@@ -120,9 +120,8 @@ func (r *vertexReconciler) reconcile(ctx context.Context, vertex *dfv1.Vertex) (
 
 	desiredReplicas := vertex.GetReplicas()
 
-	// Create PVC if needed for reduce vertices
 	if vertex.IsReduceUDF() {
-		if x := vertex.Spec.UDF.GroupBy.Storage; x != nil && x.PersistentVolumeClaim != nil {
+		if x := vertex.Spec.UDF.GroupBy.Storage; x.PersistentVolumeClaim != nil {
```
Comment (whynowy) [inline pkg/reconciler/vertex/controller.go:124]: It's no harm to keep the `x != nil` check.

File: pkg/reconciler/vertex/controller.go [hunk 2]
```diff
@@ -356,11 +352,18 @@ func (r *vertexReconciler) buildPodSpec(vertex *dfv1.Vertex, pl *dfv1.Pipeline,
 					},
 				},
 			})
-			podSpec.Containers[0].VolumeMounts = append(podSpec.Containers[0].VolumeMounts, corev1.VolumeMount{
-				Name:      volName,
-				MountPath: dfv1.PathPBQMount,
+		}
+		// Add emptyDir for reduce vertex pods
+		if storage.EmptyDir != nil {
```
Comment (whynowy) [inline pkg/reconciler/vertex/controller.go:357]: tiny: use `else if` to make it more readable.

File: pkg/apis/numaflow/v1alpha1/udf.go [hunk 3]
```diff
@@ -139,7 +139,8 @@ type SlidingWindow struct {
 
 // PBQStorage defines the persistence configuration for a vertex.
 type PBQStorage struct {
-	PersistentVolumeClaim *PersistenceStrategy `json:"persistentVolumeClaim,omitempty" protobuf:"bytes,1,opt,name=persistentVolumeClaim"`
+	PersistentVolumeClaim *PersistenceStrategy         `json:"persistentVolumeClaim,omitempty" protobuf:"bytes,1,opt,name=persistentVolumeClaim"`
```
Comment (whynowy) [inline pkg/apis/numaflow/v1alpha1/udf.go:142]: `// +optional`

File: pkg/apis/numaflow/v1alpha1/udf.go [hunk 4]
```diff
@@ -139,7 +139,8 @@ type SlidingWindow struct {
...
+	EmptyDir              *corev1.EmptyDirVolumeSource `json:"emptyDir,omitempty" protobuf:"bytes,2,opt,name=emptyDir"`
```
Comment (whynowy) [inline pkg/apis/numaflow/v1alpha1/udf.go:143]: `// +optional`

And remove `// +optional` from line 117.

File: pkg/reconciler/vertex/controller.go (hunk 2 above)
Comment (yhl25) [reply pkg/reconciler/vertex/controller.go]: done

File: pkg/apis/numaflow/v1alpha1/udf.go (hunk 4 above)
Comment (yhl25) [reply pkg/apis/numaflow/v1alpha1/udf.go]: done

File: pkg/apis/numaflow/v1alpha1/udf.go (hunk 3 above)
Comment (yhl25) [reply pkg/apis/numaflow/v1alpha1/udf.go]: done

**Review by whynowy — APPROVED** [review]
LGTM

**Comment by whynowy** [comment]
BTW - where is the pbq change?

**Comment by yhl25** [comment]
> BTW - where is the pbq change?

I had missed moving that change from the old pr, just updated the pr with that.

**Review by vigith — APPROVED** [review]

**Pull request merged**